### PR TITLE
[issue#295] Implement the Lookup inside Groupby

### PR DESCRIPTION
**Pull request by luochen01**
This PR implemented the lookup statements inside GroupBy. The semantics is that after performing group by, we then perform lookup operations (left outer join) on the grouped results. Thus, now we have two kinds of lookup operations, one before groupby (i.e., lookup for each record) and another after group by (i.e., lookup for each group).

The main changes are as follows:
1. Modified `Query` to add lookup statements inside groups.
2. Modified `SQLPPGenerator` to generate code for lookups inside groups.
3. Refactored `SQLPPGenerator` to incrementally build a query (previously we generate query parts, which are composed together in the end). This scheme is more suitable for generating nested queries.
4. Modified `JSONParser` to support the syntax changes.
5. Added corresponding test cases `SQLPPGenerator` and `JSONParser` regarding this change.

### CODE REVIEW

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala [hunk 1]
```diff
@@ -264,7 +279,8 @@ class SQLPPGenerator extends AsterixQueryGenerator {
           } else {
             s"select value $sourceVar"
           }
-        ParsedResult(Seq(projectStr, "", "", ""), exprMap)
+        queryBuilder.insert(0, projectStr + "\n")
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala]: what's the purpose of this `insert`?

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala [hunk 2]
```diff
@@ -334,6 +354,13 @@ class SQLPPGenerator extends AsterixQueryGenerator {
       case _ => aggFuncExpr(typeImpl.getAggregateStr(aggr.func))
     }
   }
+
+  protected def append(queryBuilder: StringBuilder, queryStr: String): Unit = {
+    if (!queryStr.isEmpty) {
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala]: why it won't do anything if `isEmpty`?

File: zion/src/test/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGeneratorTest.scala [hunk 3]
```diff
@@ -541,8 +545,116 @@ class SQLPPGeneratorTest extends Specification {
...
+        """
+          |select tt.`state` as `state`,tt.`count` as `count`,ll0.`population` as `population`
+          |from (
+          |select `state` as `state`,coll_count(g) as `count`
+          |from twitter.ds_tweet t
+          |where similarity_jaccard(word_tokens(t.`text`), word_tokens('zika')) > 0.0
+          |and contains(t.`text`, "virus")
+          |group by t.geo_tag.stateID as `state` group as g
+          |) tt
+          |left outer join twitter.US_population ll0 on ll0.`stateID` = tt.`state`;""".stripMargin.trim
```
Comment (JavierJia) [inline zion/src/test/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGeneratorTest.scala]: nice!

File: zion/src/test/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGeneratorTest.scala [hunk 4]
```diff
@@ -541,8 +545,116 @@ class SQLPPGeneratorTest extends Specification {
...
+          |from twitter.ds_tweet t
+          |left outer join twitter.US_population l0 on l0.`stateID` = t.`geo_tag`.`stateID`
+          |where similarity_jaccard(word_tokens(t.`text`), word_tokens('zika')) > 0.0
+          |and contains(t.`text`, "virus")
+          |group by t.geo_tag.stateID as `state` group as g
+          |) tt
+          |left outer join twitter.US_literacy ll0 on ll0.`stateID` = tt.`state`
+          |limit 0
+          |offset 0;""".stripMargin.trim
+      )
```
Comment (JavierJia) [inline zion/src/test/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGeneratorTest.scala]: so cool!

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala (hunk 1 above)
Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala]: Here queryBuilder is used to incrementally build the query. `projectStr` is actually a select statement, and thus has to be appeared in the beginning of the query.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala (hunk 2 above)
Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala]: If `queryStr` is empty, it means the corresponding statement is empty as well. For example, an empty filter:Seq[FilterStatement] or unnest:Seq[UnnestStatement].

Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala]: then let's make a good name like `appendIfQueryIsNotEmpty` or something else that describe the behavior?

Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala]: And add a comment to this function explain why we don't need to append. IMO it's more like a trick to write less code. :-)

Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala]: Make sense. I used the name `appendIfNotEmpty` to describe this behavior.

File: zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala [hunk 5]
```diff
@@ -355,6 +355,11 @@ class SQLPPGenerator extends AsterixQueryGenerator {
     }
   }
 
+  /**
+    * Append a new line and queryStr to the queryBuilder if queryStr is not empty
```
Comment (JavierJia) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala]: we don't need to repeat the function name, right? 
it's better to specify the behind-scene reason  about why we don't append if the querystring is empty?

Comment (luochen01) [inline zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SQLPPGenerator.scala]: Oh I see, just added one more explanation on this.

File: zion/src/test/scala/edu/uci/ics/cloudberry/zion/model/impl/TestQuery.scala [hunk 6]
```diff
@@ -738,6 +755,178 @@ object TestQuery {
      """.stripMargin
   )
 
+  val groupLookupJSON = Json.parse(
```
Comment (JavierJia) [inline zion/src/test/scala/edu/uci/ics/cloudberry/zion/model/impl/TestQuery.scala]: @HotLemonJuice Here is the example of using Lookup inside the groupby. 😄

**Comment by JavierJia** [comment]
@HotLemonJuice you should take a look at the JSON request exmaple about `groupbyLookup` in the test case and write the similar query in the front-end.  Ask question to @luochen01 if you have any. 😄

**Comment by chenlica** [comment]
Super!  I would like to have a F2F discussion about how you implemented it.  Please ping me to discuss.  Great job!

**Pull request merged**